rfcs/text/0003-web-rpc-hooks.md

```diff
@@ -0,0 +1,280 @@
+building block into other APIs. 
+
+```js
+import {useRPC} from 'fusion-plugin-rpc';
+
+function Component() {
+  const rpc = useRPC();
+  rpc.getUser(args); // returns a promise
+}
+```
+
+This can be implemented either via a js Proxy or by serializing the supported methods on the server into the html.
+
+## `useRPCRedux`
+
+This can be built using the `useRPC` hook and the `useDispatch` hook. This provides an abstraction that should be easy to use and 
+understand, but does require a small amount of boilerplate. It is designed to be used along side the useSelector API. 
+
+```js
+import {useRPCRedux} from 'fusion-plugin-rpc-redux-react';
+import {useDispatch, useSelector} from 'react-redux'; 
+
+function Component() {
+  const rpc = useRPCRedux();
+  const user = useSelector(state => state.user.user);
+  if (!user) {
+    throw rpc.getUser();
+  }
+  return <div>Hello {user.name}</div>
+}
+```
+
+It also supports doing multiple calls in parallel.
+
+```js
+import {useRPCRedux} from 'fusion-plugin-rpc-redux-react';
+import {useDispatch, useSelector} from 'react-redux'; 
+
+function Component() {
+  const rpc = useRPCRedux();
+  const user = useSelector(state => state.user.user);
+  if (!user) {
+    throw Promise.all(rpc.getUser(), rpc.getTrips());
+  }
+  return <div>Hello {user.name}</div>
+}
+```
+
+The drawback of this API is it requires users to separately check the cached state via useSelector. This is a few extra lines of code,
+however I think it is pretty straight forward.
+
+## `createRPCResource` (experimental)
+
+This API is experimental and should likely not be implemented until further knowledge of suspense data fetching patterns is available. 
+This pattern is inspired by the logic in the react-cache package, which shows some hints of how suspense based data fetching may look.
+
+A major benefit of this approach is it provides a unified approach for managing cached state, error states, and loading states. A drawback 
+to this approach is it takes more effort to get static typing.
+
+```js
+import {createRPCResource} from 'fusion-plugin-rpc-redux-react';
+
+const useUser = createRPCResource('getUser', state => state.user.user);
+
+function Component() {
+  const user = useUser(args);
+  return <div>Hello {user.name}</div>
+}
+```
+
+## `createRPCReactorResource` (very experimental)
+
+This API is very experimental, and should likely not be implemented in the short term. It is here mostly as a thought exercise. It tries
+to provide a way to colocate all the state management and data fetching code for an RPC call into a single API. It uses the reactor pattern,
+which was created as a way to code split reducers. See https://github.com/ganemone/redux-reactors
+
+```js
+import {createRPCReactorResource} from 'fusion-plugin-rpc-redux-react';
+
+// This would be exported from a separate file
+const useUser = createRPCResource({
+  rpcId: 'getUser',
+  selector: state => state.user.user,
+  reactor: {
```

Comment:
My understanding with React Suspense is that typically error/loading cases is handled at the ancestor `<Suspense>` component. That paradigm seems a bit at odds with this API. Any thoughts on this?

Reply:
This API would still allow for that, but it also gives you a way to handle updating your state accordingly. It might be an edge case, but you could have a situation where you want to notify other parts of your app of the loading and error state without unmounting them (which would be required to have them use the Suspense and ErrorBoundary patterns)